MRN-Code/coinstac
# Redesign - Basic Computation Registry and Database Functionality

* **Problem:**
   1. The redesign requires that a new computation registry be built that can interact with the requirements put on it by the new pipeline process. Furthermore, a new database solution is needed such that users can submit new computation schema and the application can retrieve those computations that are whitelisted.
* **Solution:**
   1. `packages/coinstac-computation-registry-new` is a new computation registry that can talk to the `docker-manager` in order to pull down pipeline computations.
   1. `packages/coinstac-ui` - As the pipeline is still under development, a temporary navigation item called `Feature Test` has been created in the UI to exhibit `docker pull`s and their output. Additionally, this screen can be used to test pulling computation schema from the database. A `Submit Computation` navigation item has also been created in the UI to allow users to upload computation JSON schema into the database.
   1. `packages/coinstac-graphql` - A Hapi-based GraphQL API has been created in front of a RethinkDB storage solution to allow reading and writing of computation schema.
   1. `apollo-server` and `apollo-client` are being used to make interacting with GraphQL easier. 
* **Tests:**
   1. Follow the steps outlined in `packages/coinstac-graphql/README.md` to setup your development environment with RethinkDB, start the GraphQL API, and insert a couple computations into the new database.
   1. Follow the steps outlined in `packages/coinstac-computation-registry-new/README.md` to add it as a dependency to `server-core` and `client-core`.
   1. Start your services in the following order: `rethinkdb`, `coinstac-graphql`, `tmux setup`.
   1. Open the UI and navigate to `Submit Computation`. Use the **Add Computation Schema** button to upload the schema files, one at a time, from `packages/coinstac-graphql`. Be sure to hit **Submit** after adding them.
      1. In the future, there will be a validation step in order to unlock the **Submit** button.
   1. Click the `Feature Test` navigation item. You should see the uploaded computation schema items listed at the top of the screen.
   1. Click the **Get IO** button by either computation to display its IO schema.
   1. Insert the name of between 1 - 3 docker images into the inputs in the middle of the page. Be sure to include the tags, eg "mysql**:latest**" to avoid downloading multiple versions.
   1. Hit the **Pull Images** button to start downloading and observe the output from the operation below.
   1. Close down your `tmuxinator` setup
      1. Clear all docker images `docker rmi $(docker images -a -q)`
      1. Restart `tmuxinator` and watch server output  as it downloads "whitelisted" computations.
      1. **Note:** All computations are currently "whitelisted". This will change soon.
